client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/Action.java

```diff
@@ -18,11 +15,17 @@ public class Action implements Serializable {
      * This info can be used for different kinds of taint analysis, eg
      * to check how such values are used in the SUT
      */
-    private Set<String> inputVariables;
+    private final Set<String> inputVariables;
 
-    public Action(int index, Collection<String> inputVariables) {
+    /**
+     * A map of hostname and WireMock IP to mock external service calls.
+     */
+    private final Map<String, String> externalServiceMapping;
+
+    public Action(int index, Collection<String> inputVariables, Map<String, String> externalServiceMapping) {
         this.index = index;
         this.inputVariables = Collections.unmodifiableSet(new HashSet<>(inputVariables));
+        this.externalServiceMapping = externalServiceMapping;
```

Comment:
should use `Collections.unmodifiable` to make it immutable, like `inputVariables`